### ISSUE TITLE
What produces the Finalizer in the ETCDsnapshotfiles and which resource deletes it?

### ISSUE THREAD

**Issue by olgaglaz** [rancher/rancher#44603]
**Rancher Server Setup**
- Rancher version: 2.7.6
- Installation option: Helm Chart
- Proxy/Cert Details: we use proxy

**Information about the Cluster**
- Kubernetes version: v1.26.12+rke2r1
- Cluster Type: Local

**User Information**
- What is the role of the user logged in? Admin
  
**Describe the bug**

When Rancher is deployed etcd snapshots are creating every 12 hours by default on the local cluster. It's also possible to configure snapshots via config.yaml. They belong to the api resource ETCDSnapshottfile. 

kub6:~ # kubectl get ETCDSnapshotfile -A
NAME                                         SNAPSHOTNAME                    NODE   LOCATION                                                                         SIZE       CREATIONTIME

local-etcd-snapshot-kub6-1709031605-d2e33a   etcd-snapshot-kub6-1709031605   kub6   file:///var/lib/rancher/rke2/server/db/snapshots/etcd-snapshot-kub6-1709031605   55681056   2024-02-27T11:00:05Z

local-etcd-snapshot-kub6-1709074805-2163ed   etcd-snapshot-kub6-1709074805   kub6   file:///var/lib/rancher/rke2/server/db/snapshots/etcd-snapshot-kub6-1709074805   59449376   2024-02-27T23:00:05Z

local-etcd-snapshot-kub6-1709118004-be242a   etcd-snapshot-kub6-1709118004   kub6   file:///var/lib/rancher/rke2/server/db/snapshots/etcd-snapshot-kub6-1709118004   65196064   2024-02-28T11:00:04Z

Here the contect of one of the snapshot's yaml file:

apiVersion: k3s.cattle.io/v1
kind: ETCDSnapshotFile
metadata:
  annotations:
    etcd.rke2.cattle.io/snapshot-token-hash: f288b2239a56
  creationTimestamp: "2024-02-28T11:00:04Z"
  finalizers:
  - wrangler.cattle.io/managed-etcd-snapshots-controller
  
It includes the Finalizer. 

If we try to delete an ETCDSnapshottfile via kubectl delete ETCDSnapshots <name> --namespace fleet-default, sometimes it is deleted and sometimes not because of the Finalizer.

In case it's not deleted the yaml file ishows deletionTimeStamp:

apiVersion: k3s.cattle.io/v1
kind: ETCDSnapshotFile
metadata:
  annotations:
    etcd.rke2.cattle.io/snapshot-token-hash: f288b2239a56
  creationTimestamp: "2024-02-28T11:00:04Z"
deletionTimeStamp: "2024-02-28T11:25:00
  finalizers:
  - wrangler.cattle.io/managed-etcd-snapshots-controller

The issue is hard ro reproduce and so to test because in the most cases the snapshot file is deleted.

What produces the Finalizer in the ETCDsnapshotfiles and which resource deletes it?

**To Reproduce**
- check the ETCDSnapshotfiles on the local cluster via kubectl get ETCDSnapshotfile -A
- delete one of them via kubectl delete ETCDSnapshots <name> --namespace fleet-default
- check the yaml file if it was not deleted via  kubectl get ETCDSnapshotfile local-etcd-snapshot-kub6-1709118004-be242a -o yaml
- in order the snapshots are created quicker, create the config.yaml with this content: 
etcd-snapshot-schedule-cron:
  - "0/3 * * * *"
etcd-snapshot-retention:
  - 10
restart the service via systemctl restart rke2-server.service

**Result**
The snapshot file is mostly deleted

**Expected Result**
ETCDnapshotfile should always not be deleted because of the Finalizer.

**Comment by carlstlaurent** [comment]
I also experienced the same bug on Rancher v2.7.9, v2.7.10 and v2.8.3. 
A lot of 0B snapshots are produced and in ETCDSnapshotfile most of snapshots there are Scheduled for deletion with the finalizer but locally on each etcd node the snapshot files appear to be ok date, size and schedule cron are all ok. 
The rke2-snapshot-controller restarts multiple times a day but I haven't found anything in the logs.

**Comment by olgaglaz** [comment]
> I also experienced the same bug on Rancher v2.7.9, v2.7.10 and v2.8.3. A lot of 0B snapshots are produced and in ETCDSnapshotfile most of snapshots there are Scheduled for deletion with the finalizer but locally on each etcd node the snapshot files appear to be ok date, size and schedule cron are all ok. The rke2-snapshot-controller restarts multiple times a day but I haven't found anything in the logs.

do you see 0B snapshots on Rancher UI? and how do you manage to delete them?

**Comment by olgaglaz** [comment]
> I didn't find any way to get rid of them, they still appear in each cluster's snapshots list even after I removed the finalizer in ETCDSnapshotfile. They don't appear in rke2-etcd-snapshots config map either. I suspect it is maybe related with rke2-etcd-snapshot-extra-metadata.
> 
> [Issue 5008](https://github.com/rancher/rke2/issues/5008)
> 
> ![image](https://private-user-images.githubusercontent.com/90282214/324010314-82dd7489-039f-4f81-8f53-ec556958fe9a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM1NDA1MDQsIm5iZiI6MTcxMzU0MDIwNCwicGF0aCI6Ii85MDI4MjIxNC8zMjQwMTAzMTQtODJkZDc0ODktMDM5Zi00ZjgxLThmNTMtZWM1NTY5NThmZTlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE5VDE1MjMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwZGE1OTc3YzFhODViYzU5ODUyNDhiYzkxNTkyMDVkMTNmOWUyMzQ5ODgxNmU2ZDhlZWQwNThlNmVjYWRiODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.yPVSviA1oFEs9ZyrbI8juIy-hYLESEIpxVZGOuxt1NQ)

Yes it looks like this. I faced such issue with ETCDSnapshots of a downstream clusters (not ETCDSnapshotfiles) and didn't find any solution. ETCDSnapshots do not have the Finilazers. In case of ETCDSnapshotfiles it was necessary to restart the rke2 service via command line.
